src/pipeline/srv_pipeline.c

```diff
@@ -0,0 +1,887 @@
+			 * out_iov is way too small. No empty iovs are skipped in the sgl.
+			 */
+			D_ERROR("iov has no space available for data value\n");
+			return -DER_REC2BIG;
+		}
+		/**
+		 * No space left in current out_iov to copy the iov, and data values are never split
+		 * among different iovs. Trying one more time with the next iov in the sgl.
+		 */
+		iov_idx_bk++;
+		if (iov_idx_bk == sgl->sg_nr) {
+			D_ERROR("Consumed all iovs in the sgl, no space available for the rest of "
+				"the data\n");
+			return -DER_REC2BIG;
+		}
+		out_iov = &sgl->sg_iovs[iov_idx_bk];
+		goto try_out_iov;
+	}
+	*iov_idx = iov_idx_bk;
+
+	ptr               = out_iov->iov_buf;
+	ptr              += out_iov->iov_len;
+	memcpy(ptr, iov->iov_buf, iov->iov_len);
+
+	out_iov->iov_len += iov->iov_len;
+	if (out_iov->iov_len == iov->iov_len) {
+		/**
+		 * out_iov was a fresh new iov, so we need to increase the number of valid output
+		 * iovs in the sgl.
+		 */
+		sgl->sg_nr_out++;
+	}
+
+	return 0;
+}
+
+static int
+pack_record(d_iov_t *d_key_iter, daos_iod_t *iods_iter, d_sg_list_t *sgl_recx_iter,
+	    uint32_t dkey_idx, struct pack_ret_data_args *pack_args)
+{
+	uint32_t idx;
+	uint32_t i;
+	uint32_t nr_iods = pack_args->nr_iods;
+	int      rc;
+
+	/**
+	 * dkey
+	 */
+	idx = pack_args->keys_iov_idx;
+	rc  = pack_value(pack_args->keys, &idx, d_key_iter);
+	if (rc != 0)
+		return rc;
+	pack_args->keys_iov_idx             = idx;
+	pack_args->kds[dkey_idx].kd_key_len = d_key_iter->iov_len;
+
+	/**
+	 * akeys' record data
+	 */
+	idx = pack_args->recx_iov_idx;
+	for (i = 0; i < nr_iods; i++) {
+		rc = pack_value(pack_args->recx, &idx, sgl_recx_iter[i].sg_iovs);
+		if (rc != 0)
+			return rc;
+
+		pack_args->recx_iov_idx                      = idx;
+		pack_args->recx_size[dkey_idx * nr_iods + i] = iods_iter[i].iod_size;
+	}
+
+	return 0;
+}
+
+/** TODO: This code still assumes dkey==NULL. The code for dkey!=NULL has to be written */
+static int
+ds_pipeline_run(daos_handle_t vos_coh, daos_unit_oid_t oid, daos_pipeline_t pipeline,
+		daos_epoch_range_t epr, uint64_t flags, daos_key_t *dkey, uint32_t nr_iods,
+		uint32_t *nr_iods_out, daos_iod_t *iods, daos_anchor_t *anchor, uint32_t nr_kds,
+		uint32_t *nr_kds_out, daos_key_desc_t *kds, daos_size_t *recx_size,
+		d_sg_list_t *sgl_keys, d_sg_list_t *sgl_recx, d_sg_list_t *sgl_agg,
+		daos_pipeline_stats_t *stats)
+{
+	int                         rc;
+	uint32_t                    nr_kds_pass;
+	d_iov_t                     d_key_iter;
+	d_sg_list_t                *sgl_recx_iter      = NULL;
+	daos_iod_t                 *iods_iter          = NULL;
+	struct enum_credits         credits            = {0};
+	struct vos_iter_anchors     anchors            = {0};
+	struct pipeline_compiled_t  pipeline_compiled  = {0};
+	struct filter_part_run_t    pipe_run_args      = {0};
+	struct pack_ret_data_args   pack_args          = {0};
+
+	*nr_kds_out  = 0;
+	*nr_iods_out = 0;
+	rc           = d_pipeline_check(&pipeline);
```

Comment:
input check? probably should be done on client side before packing RPC?

Reply:
@wangdi1 Good point. I can move it to the client.

Reply:
@wangdi1 Actually, we should not, for security reasons. But we can check it twice, in client first and then on server to discard malicious users.